source/SpinalHDL/Libraries/Pipeline/introduction.rst

```diff
@@ -0,0 +1,817 @@
+
+    n0.driveFrom(up)((self, payload) => self(IN) := payload)
+    n2.driveTo(down)((payload, self) => payload := self(OUT))
+
+
+In order to reduce verbosity, there is a set of implicit conversions between Payload toward their data representation which can be used when you are in the context of a Node : 
+
+.. code-block:: scala
+
+    val VALUE = Payload(UInt(16 bits))
+    val n1 = new Node{
+        val PLUS_ONE = insert(VALUE + 1) // VALUE is implicitly converted into its n1(VALUE) representation
+    }
+
+You can also use those implicit conversions by importing them : 
+
+.. code-block:: scala
+
+    val VALUE = Payload(UInt(16 bits))
+    val n1 = Node()
+
+    val n1Stuff = new Area {
+        import n1._
+        val PLUS_ONE = insert(VALUE) + 1 // Equivalent to n1.insert(n1(VALUE)) + 1
+    }
+
+
+There is also an API which alows you to create new Area which provide the whole API of a given node instance (including implicit convertion) without import : 
+
+.. code-block:: scala
+
+    val n1 = Node()
+    val VALUE = Payload(UInt(16 bits))
+
+    val n1Stuff = new n1.Area{
+        val PLUS_ONE = insert(VALUE) + 1 // Equivalent to n1.insert(n1(VALUE)) + 1
+    }
+
+Such feature is very useful when you have parametrizable pipeline locations for your hardware (see retiming example).
+
+
+Links
+============
+
+There is few different Links already implemented (but you could also create your own custom one).
+The idea of Links is to connect two nodes together in various ways.
+They generally have a `up` Node and a `down` Node.
+
+DirectLink
+------------------
+
+Very simple, it connect two nodes with wires only. Here is an example : 
+
+
+.. code-block:: scala
+    
+    val c01 = DirectLink(n0, n1)
+
+
+
+StageLink
+------------------
+
+This connect two nodes using registers on the data / valid signals and some arbitration on the ready.
+
+.. code-block:: scala
+    
+    val c01 = StageLink(n0, n1)
+
+
+S2mLink
+------------------
+
+This connect two nodes using registers on the ready signal, which can be useful to improve backpresure combinatorial timings.
+
+.. code-block:: scala
+    
+    val c01 = S2mLink(n0, n1)
+
+CtrlLink
+------------------
+
+This is kind of a special Link, as connect two nodes with optional flow control / bypass logic. Its API should be flexible enough to implement a CPU stage with it.
+
+Here is its flow control API (The Bool arguments enable the features) :
+
+.. list-table::
+   :header-rows: 1
+   :widths: 2 5
+
+   * - API
+     - Description
+   * - haltWhen(Bool)
+     - Allows to block the current transaction (clear up.ready down.valid)
+   * - throwWhen(Bool)
+     - Allows to cancel the current transaction from the pipeline (clear down.valid and c the transaction driver)
+   * - forgetOneWhen(Bool)
+     - Allows to request the upstream to forget the current transaction  (but doesn't clear the down.valid)
+   * - ignoreReadyWhen(Bool)
```

Comment:
forgetOneWhen ?
It will not hang anything, just make the upstream transaction driver "forget" about its current state.
That's a light weight to flush / remove something from the pipeline without having to manipulate the ready wire

Reply:
> forgetOneWhen ? It will not hang anything, just make the upstream transaction driver "forget" about its current state. That's a light weight to flush / remove something from the pipeline without having to manipulate the ready wire

i mean the ignoreReadyWhen

Reply:
Ahhh. it is to fire the upstream.
It doesn't hang downstream

Its usage is to implement some logic to flush / remove elements of a pipeline.
You can do that for instance by combining ignoreReadyWhen and terminateWhen